File: src/game/defense/defenseTransport.ts
/**
 * Defense mode musical transport: bar boundary math on audio clock seconds.
 */

export const barSeconds = (bpm: number, beatsPerBar: number): number => {
  const safeBpm = Math.max(1, bpm);
  const safeBeats = Math.max(1, beatsPerBar);
  return (60 / safeBpm) * safeBeats;
};

export const beatSeconds = (bpm: number, playbackRatio = 1): number => {
  const safeBpm = Math.max(1, bpm);
  const safeRatio = Math.max(0.0001, playbackRatio);
  return 60 / safeBpm / safeRatio;
};

export const barSecondsFromLoop = (
  loopStartSec: number,
  loopEndSec: number,
  barCount: number,
): number => {
  const duration = Math.max(1e-6, loopEndSec - loopStartSec);
  return duration / Math.max(1, Math.trunc(barCount));
};

/**
 * Keep the current bar-phase when tempo changes, so the next bar head
 * stays musically aligned instead of being recomputed from t=0.
 */
export const rebaseTransportStart = (
  now: number,
  transportStart: number,
  oldBarSec: number,
  newBarSec: number,
): number => {
  if (oldBarSec <= 0 || newBarSec <= 0) {
    return now;
  }
  const elapsed = Math.max(0, now - transportStart);
  const barIndex = Math.floor(elapsed / oldBarSec);
  const phase = elapsed - barIndex * oldBarSec;
  const fraction = phase / oldBarSec;
  return now - fraction * newBarSec;
};

export interface DefenseSwitchPlan {
  readonly switchAt: number;
  readonly immediate: boolean;
  readonly cutAt: number;
}

/** Minimum time before a scheduled cut; below this, fall back to immediate switch. */
const SCHEDULE_MIN_QUANTUM_SEC = 0.001;

/**
 * Production phrase switch timing:
 * - before cut: always target the upcoming cut (even inside the old 100ms lead window)
 * - within one beat after cut: immediate switch
 * - more than one beat after cut: target the next cut
 */
export const planDefenseSwitch = (params: {
  readonly now: number;
  readonly transportStart: number;
  readonly cutIntervalSec: number;
  readonly beatSec: number;
}): DefenseSwitchPlan => {
  const { now, transportStart, cutIntervalSec, beatSec } = params;
  if (cutIntervalSec <= 0) {
    return { switchAt: now, immediate: true, cutAt: now };
  }

  const safeBeat = Math.max(1e-9, beatSec);
  const epsilon = 1e-9;

  const elapsed = Math.max(0, now - transportStart);
  const cutIndex = Math.floor(elapsed / cutIntervalSec + epsilon);
  const recentCutAt = transportStart + cutIndex * cutIntervalSec;
  const upcomingCutAt = recentCutAt + cutIntervalSec;

  if (now >= recentCutAt - epsilon) {
    const overshoot = now - recentCutAt;
    if (overshoot <= safeBeat + epsilon) {
      return { switchAt: now, immediate: true, cutAt: recentCutAt };
    }
  }

  if (now < upcomingCutAt - epsilon) {
    const remaining = upcomingCutAt - now;
    if (remaining < SCHEDULE_MIN_QUANTUM_SEC) {
      return { switchAt: now, immediate: true, cutAt: upcomingCutAt };
    }
    return { switchAt: upcomingCutAt, immediate: false, cutAt: upcomingCutAt };
  }

  return { switchAt: now, immediate: true, cutAt: upcomingCutAt };
};

export const nextSwitchTime = (
  now: number,
  transportStart: number,
  barSec: number,
  _deadlineSec: number,
  beatSec?: number,
): number => (
  planDefenseSwitch({
    now,
    transportStart,
    cutIntervalSec: barSec,
    beatSec: beatSec ?? barSec / 4,
  }).switchAt
);

export const scheduleDeadlineSec = (baseLatencySec: number): number => (
  0.1 + Math.max(0, baseLatencySec)
);

File: ios/Jazzify/Defense/Engine/DefenseTransport.swift
import Foundation

struct DefenseSwitchPlan: Equatable {
    let switchAt: Double
    let immediate: Bool
    let cutAt: Double
}

enum DefenseTransport {
    static func barSeconds(bpm: Double, beatsPerBar: Int) -> Double {
        let safeBpm = max(1.0, bpm)
        let safeBeats = max(1, beatsPerBar)
        return (60.0 / safeBpm) * Double(safeBeats)
    }

    static func beatSeconds(bpm: Double, playbackRatio: Double = 1) -> Double {
        let safeBpm = max(1.0, bpm)
        let safeRatio = max(0.0001, playbackRatio)
        return 60.0 / safeBpm / safeRatio
    }

    static func barSecondsFromLoop(loopStartSec: Double, loopEndSec: Double, barCount: Int) -> Double {
        let duration = max(1e-6, loopEndSec - loopStartSec)
        return duration / Double(max(1, barCount))
    }

    /// Keep the current bar-phase when tempo changes, so the next bar head
    /// stays musically aligned instead of being recomputed from t=0.
    static func rebaseTransportStart(
        now: Double,
        transportStart: Double,
        oldBarSec: Double,
        newBarSec: Double
    ) -> Double {
        guard oldBarSec > 0, newBarSec > 0 else { return now }
        let elapsed = max(0, now - transportStart)
        let barIndex = floor(elapsed / oldBarSec)
        let phase = elapsed - barIndex * oldBarSec
        let fraction = phase / oldBarSec
        return now - fraction * newBarSec
    }

    private static let scheduleMinQuantumSec = 0.001

    static func planSwitch(
        now: Double,
        transportStart: Double,
        cutIntervalSec: Double,
        beatSec: Double
    ) -> DefenseSwitchPlan {
        guard cutIntervalSec > 0 else {
            return DefenseSwitchPlan(switchAt: now, immediate: true, cutAt: now)
        }

        let safeBeat = max(1e-9, beatSec)
        let epsilon = 1e-9
        let elapsed = max(0, now - transportStart)
        let cutIndex = Int(floor(elapsed / cutIntervalSec + epsilon))
        let recentCutAt = transportStart + Double(cutIndex) * cutIntervalSec
        let upcomingCutAt = recentCutAt + cutIntervalSec

        if now >= recentCutAt - epsilon {
            let overshoot = now - recentCutAt
            if overshoot <= safeBeat + epsilon {
                return DefenseSwitchPlan(switchAt: now, immediate: true, cutAt: recentCutAt)
            }
        }

        if now < upcomingCutAt - epsilon {
            let remaining = upcomingCutAt - now
            if remaining < scheduleMinQuantumSec {
                return DefenseSwitchPlan(switchAt: now, immediate: true, cutAt: upcomingCutAt)
            }
            return DefenseSwitchPlan(switchAt: upcomingCutAt, immediate: false, cutAt: upcomingCutAt)
        }

        return DefenseSwitchPlan(switchAt: now, immediate: true, cutAt: upcomingCutAt)
    }

    static func nextSwitchTime(
        now: Double,
        transportStart: Double,
        barSec: Double,
        deadlineSec: Double,
        beatSec: Double? = nil
    ) -> Double {
        _ = deadlineSec
        let resolvedBeatSec = beatSec ?? (barSec / 4.0)
        return planSwitch(
            now: now,
            transportStart: transportStart,
            cutIntervalSec: barSec,
            beatSec: resolvedBeatSec
        ).switchAt
    }

    static func barSamples(sampleRate: Double, bpm: Double, beatsPerBar: Int) -> Int64 {
        let safeBpm = max(1.0, bpm)
        let safeBeats = max(1, beatsPerBar)
        let barSeconds = (60.0 / safeBpm) * Double(safeBeats)
        return Int64((barSeconds * sampleRate).rounded())
    }

    static func nextSwitchSample(
        transportSample: Int64,
        barSamples: Int64,
        deadlineSamples: Int64
    ) -> Int64 {
        guard barSamples > 0 else { return transportSample }
        let barIndex = transportSample / barSamples
        var switchAt = (barIndex + 1) * barSamples
        if switchAt - transportSample < deadlineSamples {
            switchAt += barSamples
        }
        return switchAt
    }

    static func deadlineSamples(sampleRate: Double, renderQuantum: Int = 512) -> Int64 {
        Int64((0.05 * sampleRate).rounded()) + Int64(renderQuantum * 2)
    }
}
